lib/fcm/fcm.dart

```diff
@@ -0,0 +1,80 @@
+import 'dart:convert';
+
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+import '../base/firebase.dart';
+import 'message.dart';
+
+/// FCM Option defined here: https://firebase.google.com/docs/cloud-messaging/send-message#rest
+/// Don't know why you should overwrite these settings but you can.
+class FCMConfig {
+  static const scheme = 'https';
+  static const host = 'fcm.googleapis.com';
+  static const method = 'POST';
+  static const bool keepAlive = false;
+  final headers = <String,String>{};
+  String project_id;
+
+  FCMConfig(this.project_id);
+
+  String get path => '/v1/projects/$project_id/messages:send';
+}
+
+class FCM {
+  /* Singleton instance */
+  static FCM _instance;
+
+  static bool get initialized => _instance != null;
+
+  static FCM initialize(
+      {Firebase firebase, FCMConfig fcmConfig}) {
+    assert(!initialized,
+    'Firestore global instance is already initialized. Do not call this twice or create a local instance via FCM()');
+
+    _instance = FCM(
+        firebase: firebase ?? Firebase.instance, fcmConfig: fcmConfig);
+    return _instance;
+  }
+
+  static FCM get instance {
+    assert(initialized,
+    "Firestore hasn't been initialized. Call Firestore.initialize() before using this global instance. Alternatively, create a local instance via Firestore() and use that.");
+
+    return _instance;
+  }
+
+  /* Instance interface */
+  final Firebase firebase;
+  final FCMConfig fcmConfig;
+
+  FCM({@required this.firebase, @required this.fcmConfig})
+      : assert(firebase != null || Firebase.initialized,
+  'Firebase global instance not initialized, run Firebase.initialize().\nAlternatively, provide a local instance via Firestore.initialize(firebase: <firebase instance>)'),
+        assert(fcmConfig != null, 'Firebase Cloud Messaging configuration is missing.');
```

Comment:
This is a bit nitpicky, but I'd appreciate a more indepth explanation like the one above. Would go a long way for other people!